File: src/main/kotlin/com/tellenn/artifacts/clients/models/ArtifactsCharacter.java
package com.tellenn.artifacts.clients.models;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.tellenn.config.Const;
import com.tellenn.utils.ItemUtils;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

@JsonIgnoreProperties(ignoreUnknown = true)
public record ArtifactsCharacter(
        String name,
        int level,
        int gold,
        int hp,
        @JsonAlias("max_hp") int maxHp,
        int x,
        int y,
        InventorySlot[] inventory,
        int cooldown,
        String skin,
        String task,
        @JsonAlias("task_type") String taskType,
        @JsonAlias("task_total") int taskTotal,
        @JsonAlias("task_progress") int taskProgress,
        @JsonAlias("mining_level") int miningLevel,
        @JsonAlias("woodcutting_level") int woodcuttingLevel,
        @JsonAlias("fishing_level") int fishingLevel,
        @JsonAlias("weaponcrafting_level") int weaponcraftingLevel,
        @JsonAlias("gearcrafting_level") int gearcraftingLevel,
        @JsonAlias("jewelrycrafting_level") int jewelrycraftingLevel,
        @JsonAlias("cooking_level") int cookingLevel,
        @JsonAlias("alchemy_level") int alchemyLevel,
        @JsonAlias("inventory_max_items") int inventoryMaxItems,
        @JsonAlias("attack_fire") int attackFire,
        @JsonAlias("attack_earth") int attackEarth,
        @JsonAlias("attack_water") int attackWater,
        @JsonAlias("attack_air") int attackAir,
        @JsonAlias("dmg_fire") int dmgFire,
        @JsonAlias("dmg_earth") int dmgEarth,
        @JsonAlias("dmg_water") int dmgWater,
        @JsonAlias("dmg_air") int dmgAir,
        @JsonAlias("res_fire") int resFire,
        @JsonAlias("res_earth") int resEarth,
        @JsonAlias("res_water") int resWater,
        @JsonAlias("res_air") int resAir,
        @JsonAlias("weapon_slot") String weaponSlot,
        @JsonAlias("rune_slot") String runeSlot,
        @JsonAlias("shield_slot") String shieldSlot,
        @JsonAlias("helmet_slot") String helmetSlot,
        @JsonAlias("body_armor_slot") String bodyArmorSlot,
        @JsonAlias("leg_armor_slot") String legArmorSlot,
        @JsonAlias("boots_slot") String bootsSlot,
        @JsonAlias("ring1_slot") String ring1Slot,
        @JsonAlias("ring2_slot") String ring2Slot,
        @JsonAlias("amulet_slot") String amuletSlot,
        @JsonAlias("artifact1_slot") String artifact1Slot,
        @JsonAlias("artifact2_slot") String artifact2Slot,
        @JsonAlias("artifact3_slot") String artifact3Slot,
        @JsonAlias("utility1_slot") String utility1Slot,
        @JsonAlias("utility1_slot_quantity") int utility1SlotQuantity,
        @JsonAlias("utility2_slot") String utility2Slot,
        @JsonAlias("utility2_slot_quantity") int utility2SlotQuantity,
        @JsonAlias("bag_slot") String bagSlot ){

    public ArtifactsCharacter(ArtifactsCharacter character, String utility1Slot, String utility2Slot) {
        this(
                character.name(),
                character.level(),
                character.gold(),
                character.hp(),
                character.maxHp(),
                character.x(),
                character.y(),
                character.inventory(),
                character.cooldown(),
                character.skin(),
                character.task(),
                character.taskType(),
                character.taskTotal(),
                character.taskProgress(),
                character.miningLevel(),
                character.woodcuttingLevel(),
                character.fishingLevel(),
                character.weaponcraftingLevel(),
                character.gearcraftingLevel(),
                character.jewelrycraftingLevel(),
                character.cookingLevel(),
                character.alchemyLevel(),
                character.inventoryMaxItems(),
                character.attackFire(),
                character.attackEarth(),
                character.attackWater(),
                character.attackAir(),
                character.dmgFire(),
                character.dmgEarth(),
                character.dmgWater(),
                character.dmgAir(),
                character.resFire(),
                character.resEarth(),
                character.resWater(),
                character.resAir(),
                character.weaponSlot(),
                character.runeSlot(),
                character.shieldSlot(),
                character.helmetSlot(),
                character.bodyArmorSlot(),
                character.legArmorSlot(),
                character.bootsSlot(),
                character.ring1Slot(),
                character.ring2Slot(),
                character.amuletSlot(),
                character.artifact1Slot(),
                character.artifact2Slot(),
                character.artifact3Slot(),
                utility1Slot,
                100,
                utility2Slot,
                100,
                character.bagSlot()
        );
    }

    public String position() {
        return "{\"x\": "+this.x+" , \"y\": "+this.y+"}";
    }

    public int getRemainingTask(){
        return taskTotal - taskProgress;
    }

    public int getLevelOf(String job) {
        return switch (job) {
            case "mining" -> miningLevel;
            case "woodcutting" -> woodcuttingLevel;
            case "fishing" -> fishingLevel;
            case "cooking" -> cookingLevel;
            case "weaponcrafting" -> weaponcraftingLevel;
            case "jewelrycrafting" -> jewelrycraftingLevel;
            case "gearcrafting" -> gearcraftingLevel;
            case "alchemy" -> alchemyLevel;
            default -> 0;
        };
    }

    /**
     * Return the lowest level ranked by 5 levels in order weapon > gear > jewel
     * @Example : Weapon 23, Gear 24, jewel 20 returns weapon
     * @Example : Weapon 23, Gear 19, jewel 20 returns gear
<     * @return the lowest skill level, floored by 5, with priority.
     */
    public String getLowestCraftLevel() {
        int modulo = 5;
        int min = Math.min(Math.min(weaponcraftingLevel - (weaponcraftingLevel % modulo), jewelrycraftingLevel - (jewelrycraftingLevel % modulo)), gearcraftingLevel - (gearcraftingLevel % modulo));
        if (min == Const.MAX_LEVEL) return "none";
        if (min == weaponcraftingLevel - (weaponcraftingLevel % modulo)) return "weaponcrafting";
        if (min == gearcraftingLevel - (gearcraftingLevel % modulo)) return "gearcrafting";
        return "jewelrycrafting";

    }

    public int getFreeInventorySlots() {
        AtomicInteger slots = new AtomicInteger(inventoryMaxItems);
        Arrays.stream(inventory).toList().forEach(e -> slots.set(slots.get() - e.quantity()));
        return slots.get();
    }

    public String get(String equipmentType) {
        return switch (equipmentType) {
            case "weapon_slot" -> weaponSlot;
            case "shield_slot" -> shieldSlot;
            case "helmet_slot" -> helmetSlot;
            case "body_armor_slot" -> bodyArmorSlot;
            case "leg_armor_slot" -> legArmorSlot;
            case "boots_slot" -> bootsSlot;
            case "ring1_slot" -> ring1Slot;
            case "ring2_slot" -> ring2Slot;
            case "amulet_slot" -> amuletSlot;
            case "artifact1_slot" -> artifact1Slot;
            case "artifact2_slot" -> artifact2Slot;
            case "artifact3_slot" -> artifact3Slot;
            default -> null;
        };
    }

    public ItemInformation getHealingItems(ItemUtils itemUtils) {
        for (InventorySlot slot : inventory) {
            if (slot.quantity() > 0) {
                ItemInformation item = itemUtils.getItemInfo(slot.code());
                if (item.effects() != null) {
                    for (Effect effect : item.effects()) {
                        if (effect.code().equals("heal")) {
                            return item;
                        }
                    }
                }
            }
        }
        return null;
    }

    public String getStrongestAttackElement() {
        int highest = 0;
        if (attackAir > 0) highest = attackAir;
        if (attackFire > 0 && highest < attackFire) highest = attackFire;
        if (attackWater > 0 && highest < attackWater) highest = attackWater;
        if (attackEarth > 0 && highest < attackEarth) highest = attackEarth;

        if (attackAir == highest) return "air";
        if (attackFire == highest) return "fire";
        if (attackWater == highest) return "water";
        return "earth";
    }

    public List<ItemInformation> getAllEquipment(ItemUtils itemUtils) {
        List<String> codes = List.of(weaponSlot, shieldSlot, helmetSlot, bodyArmorSlot, legArmorSlot, bootsSlot, ring1Slot, ring2Slot, amuletSlot, artifact1Slot, artifact2Slot, artifact3Slot);
        return codes.stream().filter(s -> s != null && !s.isEmpty() ).map(itemUtils::getItemInfo).toList();
    }
}

File: src/main/kotlin/com/tellenn/artifacts/models/ArtifactsCharacter.kt
package com.tellenn.artifacts.models

import com.fasterxml.jackson.annotation.JsonProperty
import java.time.Instant

@Suppress("unused")
class ArtifactsCharacter(
    var name: String,
    var account: String,
    var level: Int,
    var gold: Int,
    var hp: Int,
    @param:JsonProperty("max_hp") var maxHp: Int,
    var x: Int,
    var y: Int,
    @param:JsonProperty("map_id") var mapId: Int,
    var layer: String,
    var inventory: Array<InventorySlot>,
    var cooldown: Int,
    var skin: String?,
    var task: String?,
    var initiative: Int,
    var threat: Int,
    @param:JsonProperty("dmg") var dmg: Int,
    @param:JsonProperty("wisdom") var wisdom: Int,
    @param:JsonProperty("prospecting") var prospecting: Int,
    @param:JsonProperty("critical_strike") var criticalStrike: Int,
    @param:JsonProperty("speed") var speed: Int,
    @param:JsonProperty("haste") var haste: Int,
    @param:JsonProperty("xp") var xp: Int,
    @param:JsonProperty("max_xp") var maxXp: Int,
    @param:JsonProperty("task_type") var taskType: String?,
    @param:JsonProperty("task_total") var taskTotal: Int,
    @param:JsonProperty("task_progress") var taskProgress: Int,
    @param:JsonProperty("mining_xp") var miningXp: Int,
    @param:JsonProperty("mining_max_xp") var miningMaxXp: Int,
    @param:JsonProperty("mining_level") var miningLevel: Int,
    @param:JsonProperty("woodcutting_xp") var woodcuttingXp: Int,
    @param:JsonProperty("woodcutting_max_xp") var woodcuttingMaxXp: Int,
    @param:JsonProperty("woodcutting_level") var woodcuttingLevel: Int,
    @param:JsonProperty("fishing_xp") var fishingXp: Int,
    @param:JsonProperty("fishing_max_xp") var fishingMaxXp: Int,
    @param:JsonProperty("fishing_level") var fishingLevel: Int,
    @param:JsonProperty("weaponcrafting_xp") var weaponcraftingXp: Int,
    @param:JsonProperty("weaponcrafting_max_xp") var weaponcraftingMaxXp: Int,
    @param:JsonProperty("weaponcrafting_level") var weaponcraftingLevel: Int,
    @param:JsonProperty("gearcrafting_xp") var gearcraftingXp: Int,
    @param:JsonProperty("gearcrafting_max_xp") var gearcraftingMaxXp: Int,
    @param:JsonProperty("gearcrafting_level") var gearcraftingLevel: Int,
    @param:JsonProperty("jewelrycrafting_xp") var jewelrycraftingXp: Int,
    @param:JsonProperty("jewelrycrafting_max_xp") var jewelrycraftingMaxXp: Int,
    @param:JsonProperty("jewelrycrafting_level") var jewelrycraftingLevel: Int,
    @param:JsonProperty("cooking_xp") var cookingXp: Int,
    @param:JsonProperty("cooking_max_xp") var cookingMaxXp: Int,
    @param:JsonProperty("cooking_level") var cookingLevel: Int,
    @param:JsonProperty("alchemy_xp") var alchemyXp: Int,
    @param:JsonProperty("alchemy_max_xp") var alchemyMaxXp: Int,
    @param:JsonProperty("alchemy_level") var alchemyLevel: Int,
    @param:JsonProperty("inventory_max_items") var inventoryMaxItems: Int,
    @param:JsonProperty("attack_fire") var attackFire: Int,
    @param:JsonProperty("attack_earth") var attackEarth: Int,
    @param:JsonProperty("attack_water") var attackWater: Int,
    @param:JsonProperty("attack_air") var attackAir: Int,
    @param:JsonProperty("dmg_fire") var dmgFire: Int,
    @param:JsonProperty("dmg_earth") var dmgEarth: Int,
    @param:JsonProperty("dmg_water") var dmgWater: Int,
    @param:JsonProperty("dmg_air") var dmgAir: Int,
    @param:JsonProperty("res_fire") var resFire: Int,
    @param:JsonProperty("res_earth") var resEarth: Int,
    @param:JsonProperty("res_water") var resWater: Int,
    @param:JsonProperty("res_air") var resAir: Int,
    @param:JsonProperty("weapon_slot") var weaponSlot: String?,
    @param:JsonProperty("rune_slot") var runeSlot: String?,
    @param:JsonProperty("shield_slot") var shieldSlot: String?,
    @param:JsonProperty("helmet_slot") var helmetSlot: String?,
    @param:JsonProperty("body_armor_slot") var bodyArmorSlot: String?,
    @param:JsonProperty("leg_armor_slot") var legArmorSlot: String?,
    @param:JsonProperty("boots_slot") var bootsSlot: String?,
    @param:JsonProperty("ring1_slot") var ring1Slot: String?,
    @param:JsonProperty("ring2_slot") var ring2Slot: String?,
    @param:JsonProperty("amulet_slot") var amuletSlot: String?,
    @param:JsonProperty("artifact1_slot") var artifact1Slot: String?,
    @param:JsonProperty("artifact2_slot") var artifact2Slot: String?,
    @param:JsonProperty("artifact3_slot") var artifact3Slot: String?,
    @param:JsonProperty("utility1_slot") var utility1Slot: String,
    @param:JsonProperty("utility1_slot_quantity") var utility1SlotQuantity: Int,
    @param:JsonProperty("utility2_slot") var utility2Slot: String,
    @param:JsonProperty("utility2_slot_quantity") var utility2SlotQuantity: Int,
    @param:JsonProperty("bag_slot") var bagSlot: String?,
    @param:JsonProperty("cooldown_expiration") var cooldownExpiration: Instant?,
    var effects: List<Effect> = emptyList()
) {
    fun getLevelOf(job: String): Int {
        return when (job) {
            "mining" -> miningLevel
            "woodcutting" -> woodcuttingLevel
            "fishing" -> fishingLevel
            "cooking" -> cookingLevel
            "weaponcrafting" -> weaponcraftingLevel
            "jewelrycrafting" -> jewelrycraftingLevel
            "gearcrafting" -> gearcraftingLevel
            "alchemy" -> alchemyLevel
            else -> 0
        }
    }

    operator fun get(equipmentType: String): String? {
        return when (equipmentType) {
            "weapon_slot" -> weaponSlot
            "shield_slot" -> shieldSlot
            "helmet_slot" -> helmetSlot
            "body_armor_slot" -> bodyArmorSlot
            "leg_armor_slot" -> legArmorSlot
            "boots_slot" -> bootsSlot
            "ring1_slot" -> ring1Slot
            "ring2_slot" -> ring2Slot
            "amulet_slot" -> amuletSlot
            "artifact1_slot" -> artifact1Slot
            "artifact2_slot" -> artifact2Slot
            "artifact3_slot" -> artifact3Slot
            else -> null
        }
    }
}